Added cbrt() and exp2() functions in Math module

harshsingh-24 opened the pull request:
Math Module enhancement:

- Added `cbrt()` or `cube root` function associated with Math module of LPython and it's associated integration tests
- Added `exp2()` or `exponential of 2` function associated with Math module of LPython and it's associated integration tests

harshsingh-24 commented:
Can @Thirumalai-Shaktivel @Smit-create @certik check that why these integration tests are failing?

certik commented:
The error is:
```
  File "/home/runner/work/lpython/lpython/integration_tests/test_math_02.py", line 1, in <module>
    from math import (sin, cos, tan, pi, sqrt, log, log10, log2, erf, erfc, gamma,
ImportError: cannot import name 'cbrt' from 'math' (/usr/share/miniconda/envs/test/lib/python3.10/lib-dynload/math.cpython-310-x86_64-linux-gnu.so)
```
So I assume CPython 3.10 does not implement `cbrt`?

If so, we need some way to only run these tests with CPython 3.11.

harshsingh-24 commented:
> The error is:
> ```
>   File "/home/runner/work/lpython/lpython/integration_tests/test_math_02.py", line 1, in <module>
>     from math import (sin, cos, tan, pi, sqrt, log, log10, log2, erf, erfc, gamma,
> ImportError: cannot import name 'cbrt' from 'math' (/usr/share/miniconda/envs/test/lib/python3.10/lib-dynload/math.cpython-310-x86_64-linux-gnu.so)
> ```
> So I assume CPython 3.10 does not implement `cbrt`?
> 
> If so, we need some way to only run these tests with CPython 3.11.

Yes. It was implemented in Python 3.11. So, how can we do that?

certik commented:
Well, we might need to add this as an option in cmake to only run this in Python 3.11. For now, you can create a dedicated tests for this, and do not register them with cpython, only llvm. That way it works with LPython, but we skip testing it with CPython. We test it manually with CPython 3.11.

harshsingh-24 commented:
> Well, we might need to add this as an option in cmake to only run this in Python 3.11. For now, you can create a dedicated tests for this, and do not register them with cpython, only llvm. That way it works with LPython, but we skip testing it with CPython. We test it manually with CPython 3.11.



> Well, we might need to add this as an option in cmake to only run this in Python 3.11. For now, you can create a dedicated tests for this, and do not register them with cpython, only llvm. That way it works with LPython, but we skip testing it with CPython. We test it manually with CPython 3.11.



> Well, we might need to add this as an option in cmake to only run this in Python 3.11. For now, you can create a dedicated tests for this, and do not register them with cpython, only llvm. That way it works with LPython, but we skip testing it with CPython. We test it manually with CPython 3.11.

@certik As I was discussing with @Thirumalai-Shaktivel and what you suggested earlier, I implemented a new backend that support Python 3.11 and updated the settings of those integration tests with Python 3.11. New Backend name is `cpython_3_11`. Any comments on that? @Smit-create

Smit-create commented:
Yeah, I'll need to update some tester files and CI to test this which I'll do after merging this PR. This looks good to me. I'll fix the tests once this is merged. Thanks @harshsingh-24

certik commented on integration_tests/CMakeLists.txt:
```diff
@@ -330,7 +352,7 @@ RUN(NAME test_builtin_round  LABELS cpython llvm c)
 RUN(NAME test_builtin_sum    LABELS cpython llvm c)
 RUN(NAME test_math1          LABELS cpython llvm c)
-RUN(NAME test_math_02        LABELS cpython llvm)
+RUN(NAME test_math_02        LABELS cpython_3_11 llvm)
```
This will disable this for current Python, which I would not do.

Split just the functionality that needs Python 3_11 and perhaps put into one file.

Thirumalai-Shaktivel replied:
Yup, @harshsingh-24 create a separate file and add the cbrt and exp2 tests there and enable cpython_3_11 for it. Leave cpython for the rest of the tests.

Thirumalai-Shaktivel replied:
We will not add this new option `cpython_3_11`, see: https://github.com/lcompilers/lpython/pull/1594#issuecomment-1477227858

Thirumalai-Shaktivel commented:
There is a thing that worries me, can we have two versions of python in the conda environment?

Thirumalai-Shaktivel commented:
I see, we cannot have it: https://stackoverflow.com/a/67595797/15913193

> https://github.com/lcompilers/lpython/pull/1594#issuecomment-1476653729
Well, we might need to add this as an option in cmake to only run this in Python 3.11. For now, you can create a dedicated tests for this, and do not register them with cpython, only llvm. That way it works with LPython, but we skip testing it with CPython. We test it manually with CPython 3.11.

@harshsingh-24 do one thing, we will split the cbrt and exp2 into a separate file and test only the LLVM.

harshsingh-24 commented:
> I see, we cannot have it: https://stackoverflow.com/a/67595797/15913193
> 
> > [#1594 (comment)](https://github.com/lcompilers/lpython/pull/1594#issuecomment-1476653729)
> > Well, we might need to add this as an option in cmake to only run this in Python 3.11. For now, you can create a dedicated tests for this, and do not register them with cpython, only llvm. That way it works with LPython, but we skip testing it with CPython. We test it manually with CPython 3.11.
> 
> @harshsingh-24 do one thing, we will split the `cbrt` and `exp2` tests into a separate file and test only the `LLVM`. Maybe add a TODO to test with CPython

Okay. So I will create a separate file for it and and remove the cpython_3_11 backend.

harshsingh-24 commented:
I have done all the changes as asked @certik @Thirumalai-Shaktivel. Can you please review it?

harshsingh-24 commented:
I will do a git squash.

Thirumalai-Shaktivel reviewed (approved):
LGTM as well! Thank you

czgdp1807 commented on integration_tests/CMakeLists.txt:
```diff
@@ -331,6 +331,7 @@ RUN(NAME test_builtin_divmod LABELS cpython llvm c)
 RUN(NAME test_builtin_sum    LABELS cpython llvm c)
 RUN(NAME test_math1          LABELS cpython llvm c)
 RUN(NAME test_math_02        LABELS cpython llvm)
+RUN(NAME test_math_03        LABELS llvm)
```
Add a TODO that this needs to be tested with Python 3.11 in future.

harshsingh-24 replied:
I already added a todo as new issue associated with this PR.

Thirumalai-Shaktivel replied:
I think Gagan is saying to add something like this:
https://github.com/lcompilers/lpython/blob/37918fa535b587040cf2ee91e37f402b76360e74/integration_tests/CMakeLists.txt#L435

harshsingh-24 replied:
Okay. Got it.

harshsingh-24 replied:
Added.

czgdp1807 reviewed (approved):
Other than my comment rest looks okay.
